File: internal/server/providers.go
package server

import (
	"errors"
	"net/http"
	"strings"

	"github.com/adama/local-agent/internal/acp"
)

// ACP UnstableLlmProtocol values a provider apiType may take.
const (
	llmProtocolAnthropic = "anthropic"
	llmProtocolOpenAI    = "openai"
	llmProtocolAzure     = "azure"
	llmProtocolVertex    = "vertex"
	llmProtocolBedrock   = "bedrock"
)

// validLLMProtocols is the set of accepted apiType values. The SDK's Validate()
// does not check apiType, so the handler rejects unknown values with 400 rather
// than forwarding an opaque error the agent would reject with a 500.
var validLLMProtocols = map[string]bool{
	llmProtocolAnthropic: true,
	llmProtocolOpenAI:    true,
	llmProtocolAzure:     true,
	llmProtocolVertex:    true,
	llmProtocolBedrock:   true,
}

// ----------------------------------------------------------------------------
// ACP Provider management handlers (plan item P4.11)
//
// These expose the unstable ACP providers/list|set|disable methods per live
// session. Providers are agent/connection-scoped in ACP, but our architecture
// spawns one transport per session, so management operates on the session's
// live transport (lazily started if nil, mirroring SwitchModel).
//
// Routes (registered in server.go apiRoutes, requireAuth):
//   GET    /api/sessions/{id}/providers            → list
//   PUT    /api/sessions/{id}/providers/{providerId} → set
//   DELETE /api/sessions/{id}/providers/{providerId} → disable
// ----------------------------------------------------------------------------

// handleListProviders returns the agent's configurable LLM providers for a
// session. Returns 501 when the agent did not advertise the providers
// capability (ErrProvidersUnsupported), 404 when the session doesn't exist,
// and 500 on transport errors.
func (s *Server) handleListProviders(w http.ResponseWriter, r *http.Request) {
	sessionID := r.PathValue("id")
	providers, err := s.deps.ACPClient.ListProviders(r.Context(), sessionID)
	if err != nil {
		writeProviderError(w, err)
		return
	}
	writeJSON(w, http.StatusOK, providers)
}

// handleSetProvider replaces the configuration for a single provider. The
// provider id comes from the path; the body carries apiType, baseUrl, and an
// optional headers map. Returns 400 on a malformed body or missing fields,
// 501 when the agent doesn't support providers, 404 when the session is gone.
func (s *Server) handleSetProvider(w http.ResponseWriter, r *http.Request) {
	sessionID := r.PathValue("id")
	providerID := r.PathValue("providerId")
	if providerID == "" {
		writeError(w, http.StatusBadRequest, "missing provider id in path")
		return
	}

	var req struct {
		APIType string            `json:"apiType"`
		BaseURL string            `json:"baseUrl"`
		Headers map[string]string `json:"headers,omitempty"`
	}
	if err := decodeJSON(w, r, &req); err != nil {
		writeError(w, http.StatusBadRequest, "invalid request body")
		return
	}
	if req.APIType == "" {
		writeError(w, http.StatusBadRequest, "apiType is required")
		return
	}
	if !validLLMProtocols[req.APIType] {
		writeError(w, http.StatusBadRequest, "invalid apiType: must be one of anthropic, openai, azure, vertex, bedrock")
		return
	}
	if req.BaseURL == "" {
		writeError(w, http.StatusBadRequest, "baseUrl is required")
		return
	}

	if err := s.deps.ACPClient.SetProvider(r.Context(), sessionID, providerID, req.APIType, req.BaseURL, req.Headers); err != nil {
		writeProviderError(w, err)
		return
	}
	writeJSON(w, http.StatusOK, map[string]string{statusKey: "updated"})
}

// handleDisableProvider disables a provider. Refuses (400) when the provider is
// marked Required by the agent — the ACP spec forbids disabling required
// providers, so the guard is enforced here at the handler level (the client
// does not re-check). Returns 501 when the agent doesn't support providers,
// 404 when the session is gone.
func (s *Server) handleDisableProvider(w http.ResponseWriter, r *http.Request) {
	sessionID := r.PathValue("id")
	providerID := r.PathValue("providerId")
	if providerID == "" {
		writeError(w, http.StatusBadRequest, "missing provider id in path")
		return
	}

	// Required-guard: fetch the provider list and refuse if the targeted
	// provider is marked Required. This is an extra round-trip but keeps the
	// guard at the API boundary (the client method is a thin pass-through).
	providers, err := s.deps.ACPClient.ListProviders(r.Context(), sessionID)
	if err != nil {
		writeProviderError(w, err)
		return
	}
	for _, p := range providers {
		if p.ID == providerID {
			if p.Required {
				writeError(w, http.StatusBadRequest, "provider "+providerID+" is required and cannot be disabled")
				return
			}
			break
		}
	}

	if err := s.deps.ACPClient.DisableProvider(r.Context(), sessionID, providerID); err != nil {
		writeProviderError(w, err)
		return
	}
	writeJSON(w, http.StatusOK, map[string]string{statusKey: "disabled"})
}

// writeProviderError maps an ACP client error to an HTTP status. The
// ErrProvidersUnsupported sentinel yields 501 (the agent genuinely doesn't
// support the capability); other errors are treated as transport/session
// failures and surface as 500 with the underlying message.
func writeProviderError(w http.ResponseWriter, err error) {
	if errors.Is(err, acp.ErrProvidersUnsupported) {
		writeError(w, http.StatusNotImplemented, "agent does not support provider management")
		return
	}
	// Session-not-found is a client error (404), consistent with the other
	// session handlers in api.go which match the same message.
	if strings.Contains(err.Error(), "session not found") {
		writeError(w, http.StatusNotFound, err.Error())
		return
	}
	writeError(w, http.StatusInternalServerError, err.Error())
}

File: src/api/providers.rs
//! Session-scoped ACP LLM provider management handlers (Go `providers.go`).
//!
//! Unsupported capability maps to 501 with Go's handler wording; missing
//! sessions map to 404 via [`crate::interfaces::map_api_error`].

use std::collections::HashMap;

use axum::extract::rejection::JsonRejection;
use axum::extract::{Path, State};
use axum::http::StatusCode;
use axum::Json;
use serde::Deserialize;
use serde_json::{json, Value};

use crate::interfaces::{ACPClient, AppError, ProviderInfo};

use super::{app_error, decode_json_body, ApiResponseError, AppState};

/// Accepted `apiType` values for PUT (Go `validLLMProtocols`).
const VALID_LLM_PROTOCOLS: &[&str] = &["anthropic", "openai", "azure", "vertex", "bedrock"];

#[derive(Deserialize)]
#[serde(rename_all = "camelCase")]
pub(crate) struct SetProviderRequest {
    api_type: String,
    base_url: String,
    #[serde(default)]
    headers: HashMap<String, String>,
}

/// `GET /api/sessions/{id}/providers`
pub async fn list_providers(
    State(state): State<AppState>,
    Path(session_id): Path<String>,
) -> Result<Json<Vec<ProviderInfo>>, ApiResponseError> {
    state
        .acp
        .list_providers(&session_id)
        .await
        .map(Json)
        .map_err(provider_error)
}

/// `PUT /api/sessions/{id}/providers/{provider_id}`
pub async fn set_provider(
    State(state): State<AppState>,
    Path((session_id, provider_id)): Path<(String, String)>,
    body: Result<Json<SetProviderRequest>, JsonRejection>,
) -> Result<Json<Value>, ApiResponseError> {
    let Json(request) = decode_json_body(body)?;
    if provider_id.is_empty() {
        return Err(ApiResponseError::bad_request("missing provider id in path"));
    }
    if request.api_type.is_empty() {
        return Err(ApiResponseError::bad_request("apiType is required"));
    }
    if !VALID_LLM_PROTOCOLS.contains(&request.api_type.as_str()) {
        return Err(ApiResponseError::bad_request(
            "invalid apiType: must be one of anthropic, openai, azure, vertex, bedrock",
        ));
    }
    if request.base_url.is_empty() {
        return Err(ApiResponseError::bad_request("baseUrl is required"));
    }

    state
        .acp
        .set_provider(
            &session_id,
            &provider_id,
            &request.api_type,
            &request.base_url,
            request.headers,
        )
        .await
        .map_err(provider_error)?;
    Ok(Json(json!({"status": "updated"})))
}

/// `DELETE /api/sessions/{id}/providers/{provider_id}`
pub async fn disable_provider(
    State(state): State<AppState>,
    Path((session_id, provider_id)): Path<(String, String)>,
) -> Result<Json<Value>, ApiResponseError> {
    if provider_id.is_empty() {
        return Err(ApiResponseError::bad_request("missing provider id in path"));
    }

    // Required-guard: refuse before calling disable (Go handler boundary).
    let providers = state
        .acp
        .list_providers(&session_id)
        .await
        .map_err(provider_error)?;
    if providers
        .iter()
        .any(|provider| provider.id == provider_id && provider.required)
    {
        return Err(ApiResponseError::bad_request(format!(
            "provider {provider_id} is required and cannot be disabled"
        )));
    }

    state
        .acp
        .disable_provider(&session_id, &provider_id)
        .await
        .map_err(provider_error)?;
    Ok(Json(json!({"status": "disabled"})))
}

/// Map ACP provider errors to Go's HTTP statuses and messages.
fn provider_error(error: AppError) -> ApiResponseError {
    match error {
        AppError::Unsupported(_) => ApiResponseError {
            status: StatusCode::NOT_IMPLEMENTED,
            // Go remaps the capability sentinel to this exact client message.
            message: "agent does not support provider management".into(),
        },
        other => app_error(other),
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn unsupported_maps_to_501_with_go_wording() {
        let error = provider_error(AppError::unsupported(
            "agent does not support the providers capability",
        ));
        assert_eq!(error.status, StatusCode::NOT_IMPLEMENTED);
        assert_eq!(error.message, "agent does not support provider management");
    }

    #[test]
    fn not_found_maps_to_404() {
        let error = provider_error(AppError::not_found_kind("session"));
        assert_eq!(error.status, StatusCode::NOT_FOUND);
        assert_eq!(error.message, "session not found");
    }
}
